shared/src/commonMain/kotlin/com/example/tasky/agenda/data/AgendaDataSource.kt

```diff
@@ -72,21 +76,37 @@ class AgendaDataSource(
             }
         }
 
+    suspend fun updateTask(bodyString: String): ResultWrapper<Unit, BaseError> =
+        safeCall {
+            val body = Json.decodeFromString<RemoteTask>(bodyString)
+            httpClient.put(TaskPath()) {
+                setBody(body)
+            }
+        }
+
     suspend fun updateEvent(
         event: Event,
         deletedPhotoKeys: List<String>,
         isGoing: Boolean,
         photos: List<ByteArray>,
+    ): ResultWrapper<RemoteEvent, BaseError> = updateEvent(event.toUpdateEventBody(deletedPhotoKeys, isGoing), photos)
+
+    suspend fun updateEvent(bodyString: String): ResultWrapper<RemoteEvent, BaseError> =
+        updateEvent(Json.decodeFromString(bodyString), emptyList())
```

Comment:
it is related to the data saved in the local database. Since I need to save the request body in to Room, I convert it to json string. But when I get them in AgendaRepository, it is in domain layer, where I should not import the data layer model for json deserialization. So I can only convert them here.

Reply:
Why don’t you then pass the data model to the datasource functions and let the implementation handle the conversion to and from JSON? The benefit is that it is easier to work with models than passing around strings. Those functions will accept any string, regardless of whether it represents a JSON or not, but with a model, you can only pass in that model to the function.

Reply:
sorry, I does not explain my throught clearly. I want to pass the data model to the datasource functions, but the repository does not know how to convert the string to the data model since it is in the domain layer.

The flow is as below:
1. repo get histories from local datasource, which contains `apiType`, `bodyJsonString`
2. repo call functions in datasource according to `apiType`, repo does not know how to deserialize the data model, so it passes `bodyJsonString` as params

or should I solve the problem by changing the logic when saving the offline history?

Reply:
Yes, this would be a problem when you have your domain implementation in your domain layer 😅 
I believe I’ve commented on it before 🤔 But I’ll go over it again.

You are correct; the domain layer should not have any knowledge about the data and presentation layers. The **implementation** of the domain layer is another story. It should live inside your data layer. This class can know about the data types of the local and remote data sources (`Entity` and `Response` models). Especially when it comes to a Repository implementation, its job is to map between the different models of the different sources of data. 

Take the following example:

```
// domain layer
interface AgendaRepository {
  // the input is a domain model
  suspend fun createReminder(reminder: AgendaItem): EmptyResult<DataError>
}

// data layer - repository implementation
class OfflineFirstAgendaRepository(
  // ...
) {
  override suspend fun createReminder(reminder: AgendaItem): EmptyResult<DataError> {
    // fetching an Entity model
    val localResult = localDataSource.upsertAgendaItem(reminder)
    if (localResult !is Result.Success) {
      return localResult
    }
    
    // and here we fetch a Response model
    val remoteResult = remoteDataSource.postReminder(reminder)
  
    if (remoteResult is Result.Error) {
      applicationScope.launch {
        syncScheduler.scheduleSync(
          type = SyncAgendaItemsScheduler.SyncType.CreateReminder(
            reminder = reminder
          )
        )
      }.join()
      return Result.Success(Unit)
    } else {
      applicationScope.launch {
        notificationScheduler.schedule(
          data = reminder.toNotificationData()
        )
      }.join()
    }
  
    return remoteResult
  }
}
```